### PR TITLE
 Binary readers (depends on #27)

### PR DESCRIPTION
**Pull request by dtebbs**
Compatible with libzeth serialization.

### CODE REVIEW

File: src/main/java/io/BinaryCurveReader.java [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+  public G2T readG2NoThrow() {
+    try {
+      return readG2();
+    } catch (IOException e) {
+      return null;
+    }
+  }
+
+  protected BigInteger readBigInteger(final int numBytes) throws IOException {
+    final byte[] bytes = readNBytes(numBytes);
```
Comment (AntoineRondelet) [inline src/main/java/io/BinaryCurveReader.java:69]: Shouldn't `readNBytes` be the method of an `InputStream`? On which stream are you reading here? (why don't you need to do `inStream_.readNBytes()` here?)

File: src/test/java/io/BinaryTest.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import algebra.curves.barreto_lynn_scott.bls12_377.BLS12_377Fields.BLS12_377Fr;
+import java.math.BigInteger;
+import org.junit.jupiter.api.Test;
+
+public class BinaryTest {
+
+  @Test
+  public void bignumBinayTest() {
```
Comment (AntoineRondelet) [inline src/test/java/io/BinaryTest.java:12]: `bignumBinayTest` -> bignumBinaryTest

File: src/test/java/io/AssignmentReaderTest.java [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        BN254aFr.ONE, new AssignmentReader<BN254aFr, BN254aG1, BN254aG2>(binReader), 8, 8);
+  }
+
+  @Test
+  public void testAssignmentReaderRDDBN254a_2_2() throws IOException {
+    final var in = openTestFile("assignment_alt-bn128.bin");
+    final var binReader = new BN254aBinaryReader(in);
+
+    testReaderAgainstDataRDD(
+        BN254aFr.ONE, new AssignmentReader<BN254aFr, BN254aG1, BN254aG2>(binReader), 2, 2);
```
Comment (AntoineRondelet) [inline src/test/java/io/AssignmentReaderTest.java:144]: Do I understand correctly that you are reading just a subset of the assignment here? (i.e. execute `PairRDDAggregator` with only 2 batches and 2 partitions, which won't load the full assignment, right?)

File: src/main/java/io/BinaryCurveReader.java (hunk 1 above)
Comment (dtebbs) [reply src/main/java/io/BinaryCurveReader.java:69]: BinaryCurveReader extends the DataStream (which supports reading ints, longs, etc), so readNBytes can be called on this class.

File: src/test/java/io/AssignmentReaderTest.java (hunk 3 above)
Comment (dtebbs) [reply src/test/java/io/AssignmentReaderTest.java:144]: The full assignment should always be read.  `numPartitions` should only control the partitioning of the full assignment, and `batchSize` controls how many values are loaded per batch (on the assumption that the whole assignment may not fit in local memory, so much be loaded in fragments).

File: src/test/java/io/BinaryTest.java (hunk 2 above)
Comment (dtebbs) [reply src/test/java/io/BinaryTest.java]: Thanks.  Should be fixed in latest.

File: src/test/java/io/AssignmentReaderTest.java (hunk 3 above)
Comment (AntoineRondelet) [reply src/test/java/io/AssignmentReaderTest.java:143]: Ah right, of course... Got it, thanks!

File: src/main/java/io/BinaryCurveReader.java (hunk 1 above)
Comment (AntoineRondelet) [reply src/main/java/io/BinaryCurveReader.java:69]: I see, makes sense, thanks!

**Review by AntoineRondelet — APPROVED** [review]
LGTM, thanks @dtebbs

**Pull request merged**